### PR TITLE
[FLINK-22317][table] Support DROP column/constraint/watermark for ALTER TABLE statement

### PR DESCRIPTION
**Pull request by lsyldliu**
## What is the purpose of the change

Support DROP column/constraint/watermark for ALTER TABLE statement

## Brief change log

  - *Support DROP column/constraint/watermark for ALTER TABLE statement*

## Verifying this change

This change added tests and can be verified as follows:
  - *Add unit tests in SqlToOperationConverterTest*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/AlterTableDropConstraintOperation.java [hunk 1]
```diff
@@ -20,24 +20,42 @@
...
 public class AlterTableDropConstraintOperation extends AlterTableOperation {
-    private final String constraintName;
+
+    private final boolean isPrimaryKey;
+    private final @Nullable String constraintName;
 
     public AlterTableDropConstraintOperation(
-            ObjectIdentifier tableIdentifier, String constraintName) {
+            ObjectIdentifier tableIdentifier,
+            boolean isPrimaryKey,
```
Comment (luoyuxia) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/AlterTableDropConstraintOperation.java:35]: Will it better to wrap `isPrimaryKey` and `constraintName` to a class may called `Constraint`?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java [hunk 2]
```diff
@@ -117,6 +128,193 @@ public static Operation convertAddReplaceColumns(
...
+            throw new ValidationException(
+                    String.format("Table %s does not exist watermark.", tableIdentifier));
+        }
+
+        Schema.Builder builder = Schema.newBuilder();
+        // build column
+        builder.fromColumns(originSchema.getColumns());
+
+        // build primary key
+        Optional<Schema.UnresolvedPrimaryKey> originPrimaryKey = originSchema.getPrimaryKey();
```
Comment (luoyuxia) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java:172]: should we also build constraint or primary key is the only constraint?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java [hunk 3]
```diff
@@ -117,6 +128,193 @@ public static Operation convertAddReplaceColumns(
...
+        // filter the dropped column which is not in table firstly
+        List<String> toDropColumns =
+                sqlAlterTableDropColumns.getColumns().getList().stream()
+                        .map(SqlIdentifier.class::cast)
+                        .map(SqlIdentifier::getSimple)
+                        .filter(column -> originTableColumns.contains(column))
+                        .collect(Collectors.toList());
+
+        // validate column size
+        if (originTableColumns.size() == 1 && toDropColumns.size() > 0) {
```
Comment (luoyuxia) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java:207]: What if the `toDropColumns` is equal to `originTableColumns`?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java [hunk 4]
```diff
@@ -117,6 +128,193 @@ public static Operation convertAddReplaceColumns(
...
+        // build primary key
+        Optional<Schema.UnresolvedPrimaryKey> originPrimaryKey = originSchema.getPrimaryKey();
+        if (originPrimaryKey.isPresent()) {
+            List<String> originPrimaryKeyNames = originPrimaryKey.get().getColumnNames();
+            String constrainName = originPrimaryKey.get().getConstraintName();
+            List<String> newPrimaryKeyNames =
+                    originPrimaryKeyNames.stream()
+                            .filter(pkName -> !toDropColumns.contains(pkName))
+                            .collect(Collectors.toList());
+            if (newPrimaryKeyNames.size() > 0) {
```
Comment (luoyuxia) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java:250]: What the behavior of drop one column of composite primary key in other databases?

**Review by luoyuxia — COMMENTED** [review]
@lsyldliu Thanks for contribution. I left minor comments.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java (hunk 4 above)
Comment (lsyldliu) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java:250]: mysql and postgresql.Here is some scenes when drop column:

Constraint refers to this column and other column.Remove this column from Constraint.If constraint only refers to this column ,drop column and constraint together.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java [hunk 5]
```diff
@@ -117,6 +128,193 @@ public static Operation convertAddReplaceColumns(
...
+            throw new ValidationException(
+                    String.format("Table %s does not exist primary key.", tableIdentifier));
+        }
+
+        String constraintName = null;
+        if (alterTableDropConstraint.getConstraintName().isPresent()) {
+            constraintName = alterTableDropConstraint.getConstraintName().get().getSimple();
+        }
+        if (!StringUtils.isNullOrWhitespaceOnly(constraintName)
+                && !oriPrimaryKey.get().getConstraintName().equals(constraintName)) {
```
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java:149]: Yes, primary key also is constraint, it has a constraint name. we can drop primary key in two statement:

1. `DROP PIMARY KEY ` directly
2. Specify the constraint name, use `DROP CONSTRAINT xxxx`

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/AlterTableDropConstraintOperation.java (hunk 1 above)
Comment (lsyldliu) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ddl/AlterTableDropConstraintOperation.java:35]: I think it is not needed, here only two variable.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java (hunk 2 above)
Comment (lsyldliu) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java:172]: I did not understand the meaning of this sentence, can you give more context?

**Review by lsyldliu — COMMENTED** [review]
@luoyuxia Thanks for your review, I have updated it.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java (hunk 2 above)
Comment (luoyuxia) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/utils/OperationConverterUtils.java:172]: Nerver mind. As the primary key is the only constraint, it's okay to only build primary key.

**Review by luoyuxia — APPROVED** [review]
Thanks for updating. LGTM

**Comment by fsk119** [comment]
Merged in the #21571

**Pull request closed without merging**